### PR TITLE
Wait for udev

### LINKED ISSUE #4125: Unable to find a system nic while running cloud-init
# Bug report

I run "cloud-init status", I found that cloud-init failed to executel

## Steps to reproduce the problem

Reboot computes

## Environment details
- Cloud-init version: 21.4
- Operating System Distribution: openEuler

## cloud-init logs

2023-05-18 02:48:15,451 - util.py[DEBUG]: failed stage init-local
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/cloudinit/cmd/main.py", line 689, in status_wrapper
    ret = functor(name, args)
  File "/usr/lib/python3.9/site-packages/cloudinit/cmd/main.py", line 398, in main_init
    init.apply_network_config(bring_up=bring_up_interfaces)
  File "/usr/lib/python3.9/site-packages/cloudinit/stages.py", line 789, in apply_network_config
    netcfg, src = self._find_networking_config()
  File "/usr/lib/python3.9/site-packages/cloudinit/stages.py", line 740, in _find_networking_config
    if self.datasource and hasattr(self.datasource, 'network_config'):
  File "/usr/lib/python3.9/site-packages/cloudinit/sources/DataSourceConfigDrive.py", line 158, in network_config
    self._network_config = openstack.convert_net_json(
  File "/usr/lib/python3.9/site-packages/cloudinit/sources/helpers/openstack.py", line 698, in convert_net_json
    raise ValueError("Unable to find a system nic for %s" % d)
ValueError: Unable to find a system nic for {'mtu': 1500, 'type': 'physical', 'subnets': [{'type': 'static', 'netmask': '255.255.255.0', 'routes': [{'netmask': '0.0.0.0', 'network': '0.0.0.0', 'gateway': '192.168.50.1'}], 'address': '192.168.50.19', 'ipv4': True}], 'mac_address': 'fa:16:3e:7c:49:9f'}

### PR DESCRIPTION
**Pull request by rjschwei**
## Proposed Commit Message

It is possible that we outrun udev and when we try to enumerate the macs any given mac may not yet be present. If we detect the condition give udev a chance to catch up and check the system macs again before triggering an error.

Fixes #4125

## Additional Context
This mostly addresses the issue observed in #4125. There are still corner cases where this change would not totally prevent the problem. 

## Test Steps
Customer running into this condition has verified that this change addresses their issue on SLE.

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

File: tests/unittests/sources/test_configdrive.py [hunk 1]
```diff
@@ -389,6 +389,7 @@ def setUp(self):
             M_PATH + "util.find_devs_with", "m_find_devs_with", return_value=[]
         )
         self.tmp = self.tmp_dir()
+        self.allowed_subp = True
```
Comment (holmanb) [inline tests/unittests/sources/test_configdrive.py:392]: @rjschwei without this some tests failed, right?

Setting this allows these tests to call udevadm settle on the host. It would be better to avoid the udevadm dependency altogether by mocking out the new `util.udevadm_settle` call.

**Review by holmanb — COMMENTED** [review]
I don't love this solution. An interface might not yet be in the udev event queue, which means that the same thing could still happen if a device is discovered by the kernel later than this happens. 

That said, it does close the gap when it can occur and this code path doesn't run except when a non-recoverable error would otherwise occur in cloud-init, so I'm fine with this solution for now.

That said, we can't just run `udevadm settle` in unit tests - that would potentially break BSDs, alpine, etc. Unit tests shouldn't depend on the behavior and utilities available on the host system. I'll push a patch to this branch that fixes that.

**Comment by rjschwei** [comment]
> I don't love this solution. An interface might not yet be in the udev event queue, which means that the same thing could still happen if a device is discovered by the kernel later than this happens.
> 

Yes it is not a complete fix to the solution, as I eluded to in the PR comments. I agree wit the analysis in #4125 and there is no obvious and absolute "correct" path forward. 


> That said, it does close the gap when it can occur and this code path doesn't run except when a non-recoverable error would otherwise occur in cloud-init, so I'm fine with this solution for now.

Great, thanks.

File: tests/unittests/sources/test_configdrive.py (hunk 1 above)
Comment (rjschwei) [reply tests/unittests/sources/test_configdrive.py]: Thanks for fixing this for me :D
